sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py

```diff
@@ -353,6 +518,382 @@ def make_stage(base_stage, transform):
+                  spec=beam_runner_api_pb2.SdkFunctionSpec(
+                      spec=beam_runner_api_pb2.FunctionSpec(
+                          urn=common_urns.coders.KV.urn)),
+                  component_coder_ids=[key_coder_id, spec.timer_coder_id]))
+          # Inject the read and write pcollections.
+          timer_read_pcoll = unique_name(
+              pipeline_context.components.pcollections,
+              '%s_timers_to_read_%s' % (transform.unique_name, tag))
+          timer_write_pcoll = unique_name(
+              pipeline_context.components.pcollections,
+              '%s_timers_to_write_%s' % (transform.unique_name, tag))
+          pipeline_context.components.pcollections[timer_read_pcoll].CopyFrom(
+              beam_runner_api_pb2.PCollection(
+                  unique_name=timer_read_pcoll,
+                  coder_id=key_timer_coder_id,
+                  windowing_strategy_id=input_pcoll.windowing_strategy_id,
+                  is_bounded=input_pcoll.is_bounded))
+          pipeline_context.components.pcollections[timer_write_pcoll].CopyFrom(
+              beam_runner_api_pb2.PCollection(
+                  unique_name=timer_write_pcoll,
+                  coder_id=key_timer_coder_id,
+                  windowing_strategy_id=input_pcoll.windowing_strategy_id,
+                  is_bounded=input_pcoll.is_bounded))
+          stage.transforms.append(
+              beam_runner_api_pb2.PTransform(
+                  unique_name=timer_read_pcoll + '/Read',
+                  outputs={'out': timer_read_pcoll},
+                  spec=beam_runner_api_pb2.FunctionSpec(
+                      urn=bundle_processor.DATA_INPUT_URN,
+                      payload=create_buffer_id(
+                          timer_read_pcoll, kind='timers'))))
+          stage.transforms.append(
+              beam_runner_api_pb2.PTransform(
+                  unique_name=timer_write_pcoll + '/Write',
+                  inputs={'in': timer_write_pcoll},
+                  spec=beam_runner_api_pb2.FunctionSpec(
+                      urn=bundle_processor.DATA_OUTPUT_URN,
+                      payload=create_buffer_id(
+                          timer_write_pcoll, kind='timers'))))
+          assert tag not in transform.inputs
+          transform.inputs[tag] = timer_read_pcoll
+          assert tag not in transform.outputs
+          transform.outputs[tag] = timer_write_pcoll
+          stage.timer_pcollections.append(
+              (timer_read_pcoll + '/Read', timer_write_pcoll))
+    yield stage
+
+
+def sort_stages(stages, pipeline_context):
+  """Order stages suitable for sequential execution.
+  """
+  seen = set()
+  ordered = []
+
+  def process(stage):
+    if stage not in seen:
+      seen.add(stage)
+      for prev in stage.must_follow:
+        process(prev)
+      ordered.append(stage)
+  for stage in stages:
+    process(stage)
+  return ordered
+
+
+def window_pcollection_coders(stages, pipeline_context):
```

Comment:
Can we add a docstring here for consistency?

Reply:
Done.